cmd/zstream/zstream_decompress.c

```diff
@@ -0,0 +1,264 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright 2022 Axcient.  All rights reserved.
+ * Use is subject to license terms.
+ */
+
+#include <err.h>
+#include <search.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/zfs_ioctl.h>
+#include <sys/zio_checksum.h>
+#include "zfs_fletcher.h"
+#include "zstream.h"
+
+static int
+dump_record(dmu_replay_record_t *drr, void *payload, int payload_len,
+    zio_cksum_t *zc, int outfd)
+{
+	assert(offsetof(dmu_replay_record_t, drr_u.drr_checksum.drr_checksum)
+	    == sizeof (dmu_replay_record_t) - sizeof (zio_cksum_t));
+	fletcher_4_incremental_native(drr,
+	    offsetof(dmu_replay_record_t, drr_u.drr_checksum.drr_checksum), zc);
+	if (drr->drr_type != DRR_BEGIN) {
+		assert(ZIO_CHECKSUM_IS_ZERO(&drr->drr_u.
+		    drr_checksum.drr_checksum));
+		drr->drr_u.drr_checksum.drr_checksum = *zc;
+	}
+	fletcher_4_incremental_native(&drr->drr_u.drr_checksum.drr_checksum,
+	    sizeof (zio_cksum_t), zc);
+	if (write(outfd, drr, sizeof (*drr)) == -1)
+		return (errno);
+	if (payload_len != 0) {
+		fletcher_4_incremental_native(payload, payload_len, zc);
+		if (write(outfd, payload, payload_len) == -1)
+			return (errno);
+	}
+	return (0);
+}
+
+int
+zstream_do_decompress(int argc, char *argv[])
+{
+	int bufsz = SPA_MAXBLOCKSIZE;
+	char *buf = safe_malloc(bufsz);
+	dmu_replay_record_t thedrr;
+	dmu_replay_record_t *drr = &thedrr;
+	zio_cksum_t stream_cksum;
+	int i;
+
+	if (hcreate(argc) == 0)
+		errx(1, "hcreate");
+	for(i = 0; i < argc; i++) {
+		ENTRY e = {.key = argv[i]};
+		ENTRY *p;
+
+		p = hsearch(e, ENTER);
+		if (p == NULL)
+			errx(1, "hsearch");
```

Comment:
I wonder, since we don't use the data member at all, if `tsearch()` with `strcmp()`, or even `qsort()`/`bsearch()` (the latter of which is inlined by glibc) over argv may make more sense.

Also, I'm assuming the `object,offset` pair is provided verbatim and in decimal when found by the scrub (the bug doesn't show the output), but even then it may make more sense to decode this (with `sscanf()` if it's decimal, `strsep()` into `strtoull()`) into a struct and `tsearch()` or instead of string-formatting the block numbers.

Reply:
Scrub doesn't find this kind of corruption.  As far as scrub is concerned the data is just fine, because its checksums are correct.  You need another method to find it.  As far as the search goes, hsearch uses a hash table, so I would expect it to outperform a binary search tree for large datasets.  And for small ones, I like the small code size.  I'm not terribly concerned about performance, because I hope not to run this command very often!

Reply:
If performance is not an issue (but I'd expect sprintf+hash+hsearch to overtake tsearch over integers), but correctness *is*, I'd still prefer this to actually parse/validate the numbers or at least explicitly say that `inode` and `offset` are decimal in the manual.